Lab-2/jonasnylund_lab2.ipynb

```diff
@@ -0,0 +1,764 @@
+        "\n",
+        "print(m,n);\n",
+        "print(A @ x);\n",
+        "print(sparseProduct(x, A.data, A.indices, A.indptr));\n"
+      ],
+      "execution_count": 87,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "5 7\n",
+            "[0.81479742 0.         0.60141252 1.23667301 0.53705956]\n",
+            "[0.81479742 0.         0.60141252 1.23667301 0.53705956]\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "WiUzO6hhpM3a",
+        "colab_type": "text"
+      },
+      "source": [
+        "**QR factorization**\n",
+        "\n",
+        "Factorizing a matrix **A** into a orthogonal matrix **Q** and a triangular matrix **R** can be used to solve certain linear algebra equations efficiently. The factorization can be done in multiple ways, here is an implementation of the modified Gram–Schmidt process."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "vrGDAuFUpRki",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def qrFactorize(mat):\n",
+        "  A = np.array(mat);\n",
+        "  m = A.shape[0];\n",
+        "  n = A.shape[1];\n",
+        "  assert m >= n;\n",
+        "\n",
+        "  #==============================================#\n",
+        "  # https://f.kth.se/sangbok/#!/chapter/8/song/16\n",
+        "  i = 0;\n",
+        "\n",
+        "  ##Tag en delrumsbas M och en vektor a\n",
+        "  M = [];\n",
+        "  while i < n:\n",
+        "    a = np.array(A[:,i]);\n",
+        "\n",
+        "    ##Projicera ner, tag dess residual\n",
+        "    for j in range(i):\n",
+        "      a -= np.dot(A[:,i], M[j])*M[j];\n",
+        "\n",
+        "    ## Normalisera, tillför den till M\n",
+        "    M.append( a/np.linalg.norm(a) );\n",
+        "\n",
+        "    ##Ta sen nästa vektor, börja om igen\n",
+        "    i += 1;\n",
+        "  #==============================================#\n",
+        "\n",
+        "  Q = np.zeros(A.shape);\n",
+        "  R = np.zeros((n,n));\n",
+        "  for i in range(n):\n",
+        "    Q[:,i] = M[i];\n",
+        "    for j in range(i, n):\n",
+        "      R[i,j] = np.dot(A[:,j], M[i]);\n",
+        "\n",
+        "  return Q, R;\n"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "eesElruytMYp",
+        "colab_type": "code",
+        "outputId": "2e2c9cbd-d444-4661-f742-2721d6f2cdbe",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 69
+        }
+      },
+      "source": [
+        "n = np.random.randint(2,10);\n",
+        "m = np.random.randint(n,10);\n",
+        "\n",
+        "A = np.random.rand(m,n);\n",
+        "Q,R = qrFactorize(A);\n",
+        "q,r = np.linalg.qr(A);\n",
```

Comment:
You do not use these values to prove that your implementation is correct. You need to modify the tests to show correctness of your qr-factorization.

As your implementation does not follow the algorithm given by the lecture notes, it is more important to show that it is correct.

Reply:
Yes, I would like you to include tests comparing your solution to the np-solution. Otherwise it looks good.

Reply:
I have now implemented the comparison with numpy